In [1]:
credit = read.csv('../../R_projects//MLwR//Machine Learning with R (2nd Ed.)//Chapter 05//credit.csv')

In [2]:
library(caret)

Loading required package: lattice
Loading required package: ggplot2


In [3]:
modelLookup('C5.0')

model parameter label                 forReg forClass probModel
1 C5.0  trials    # Boosting Iterations FALSE  TRUE     TRUE     
2 C5.0  model     Model Type            FALSE  TRUE     TRUE     
3 C5.0  winnow    Winnow                FALSE  TRUE     TRUE

In [4]:
set.seed(300)

In [5]:
m = train(default ~ ., data = credit, method = 'C5.0')

Loading required package: C50
Loading required package: plyr


In [6]:
m

C5.0 

1000 samples
  16 predictor
   2 classes: 'no', 'yes' 

No pre-processing
Resampling: Bootstrapped (25 reps) 
Summary of sample sizes: 1000, 1000, 1000, 1000, 1000, 1000, ... 
Resampling results across tuning parameters:

  model  winnow  trials  Accuracy   Kappa    
  rules  FALSE    1      0.6847204  0.2578421
  rules  FALSE   10      0.7112829  0.3094601
  rules  FALSE   20      0.7221976  0.3260145
  rules   TRUE    1      0.6888432  0.2549192
  rules   TRUE   10      0.7113716  0.3038075
  rules   TRUE   20      0.7233222  0.3266866
  tree   FALSE    1      0.6769653  0.2285102
  tree   FALSE   10      0.7222552  0.2880662
  tree   FALSE   20      0.7297858  0.3067404
  tree    TRUE    1      0.6771020  0.2219533
  tree    TRUE   10      0.7173312  0.2777136
  tree    TRUE   20      0.7285714  0.3058474

Accuracy was used to select the optimal model using  the largest value.
The final values used for the model were trials = 20, model = tree and winnow
 = FALSE.

1. A brief description of the input dataset.
2. A report of the preprocessing and resampling methods applied
3. A list of the candidate models evaluated: 12 different models were tested based on the combinations of three C5.0 tuning parameters
4. The choice of the best model

In [7]:
p = predict(m, credit)

In [8]:
table(p, credit$default)

     
p      no yes
  no  700   2
  yes   0 298

In [9]:
head(predict(m, credit))

[1] no  yes no  no  yes no 
Levels: no yes

In [10]:
head(predict(m, credit, type = 'prob'))

no        yes       
1 0.9606970 0.03930299
2 0.1388444 0.86115561
3 1.0000000 0.00000000
4 0.7720279 0.22797208
5 0.2948062 0.70519385
6 0.8583715 0.14162851

## choosing the resampling method

trainControl() is used to create a set of configuration options known as a control object, which guides the train() function.

In [11]:
ctrl = trainControl(method = 'cv', number = 10, 
                   selectionFunction = 'oneSE')

## choosing the grid of parameters to optimize

In [12]:
grid = expand.grid(.model = 'tree',
                  .trials = c(1, 5, 10, 15, 20, 25, 30, 35),
                  .winnow = "FALSE")

In [13]:
grid

.model .trials .winnow
1 tree    1      FALSE  
2 tree    5      FALSE  
3 tree   10      FALSE  
4 tree   15      FALSE  
5 tree   20      FALSE  
6 tree   25      FALSE  
7 tree   30      FALSE  
8 tree   35      FALSE

In [14]:
set.seed(300)

In [15]:
m = train(default ~ ., data = credit, method = "C5.0",
         metric = 'Kappa',
         trControl = ctrl,
         tuneGrid = grid)

Warning message in Ops.factor(x$winnow):
“‘!’ not meaningful for factors”

In [16]:
m

C5.0 

1000 samples
  16 predictor
   2 classes: 'no', 'yes' 

No pre-processing
Resampling: Cross-Validated (10 fold) 
Summary of sample sizes: 900, 900, 900, 900, 900, 900, ... 
Resampling results across tuning parameters:

  trials  Accuracy  Kappa    
   1      0.724     0.3124461
   5      0.713     0.2921760
  10      0.719     0.2947271
  15      0.721     0.3009258
  20      0.717     0.2929875
  25      0.728     0.3150336
  30      0.729     0.3104144
  35      0.741     0.3389908

Tuning parameter 'model' was held constant at a value of tree
Tuning
 parameter 'winnow' was held constant at a value of FALSE
Kappa was used to select the optimal model using  the one SE rule.
The final values used for the model were trials = 1, model = tree and winnow
 = FALSE.

## improving model performance with meta-learning (teamwork-based techniques / ensembles)

+ better generalizability to future problems
+ improved performance on massive or miniscule datasets
+ ability to synthesize data from distinct domains
+ a more nuanced understanding of difficult learning tasks

## Bagging

Uses a technique called bootstrap aggregating or bagging. Bagging generates a number of training datasets by bootstrap sampling the original data. These datasets are then used to generate a set of models using a single learning algorithm. Predictions are combined using voting or averaging (for numeric prediction).

In [17]:
library('ipred')

In [18]:
set.seed(300)
mybag = bagging(default ~ ., data = credit, nbagg = 25)

In [19]:
credit_pred = predict(mybag, credit)

In [20]:
table(credit_pred, credit$default)

           
credit_pred  no yes
        no  699   2
        yes   1 298

In [21]:
set.seed(300)
ctrl = trainControl(method = 'cv', number = 10)
train(default ~ ., data = credit, method = 'treebag', trControl = ctrl)

Loading required package: e1071


Bagged CART 

1000 samples
  16 predictor
   2 classes: 'no', 'yes' 

No pre-processing
Resampling: Cross-Validated (10 fold) 
Summary of sample sizes: 900, 900, 900, 900, 900, 900, ... 
Resampling results:

  Accuracy  Kappa    
  0.747     0.3618416


In [22]:
str(svmBag)

List of 3
 $ fit      :function (x, y, ...)  
 $ pred     :function (object, x)  
 $ aggregate:function (x, type = "class")  


In [23]:
svmBag$fit

function (x, y, ...) 
{
    loadNamespace("kernlab")
    out <- kernlab::ksvm(as.matrix(x), y, prob.model = is.factor(y), 
        ...)
    out
}
<environment: namespace:caret>

In [24]:
bagctrl = bagControl(fit = svmBag$fit,
                    predict = svmBag$pred,
                    aggregate = svmBag$aggregate)

In [26]:
library(caret)
library(ipred)

In [28]:
set.seed(300)
svmbag <- train(default ~ ., data = credit, "bag",
                trControl = ctrl, bagControl = bagctrl)

Warning message in data.row.names(row.names, rowsi, i):
“some row.names duplicated: 3,6,8,9,10,13,15,17,22,27,28,32,33,36,37,39,42,43,45,47,50,51,55,57,59,60,63,64,65,67,69,73,78,80,86,89,95,100,104,110,112,118,120,121,124,128,129,132,134,135,138,139,140,144,150,151,154,155,157,162,165,166,167,169,175,177,179,180,182,190,194,195,201,203,207,209,213,215,216,218,220,222,225,227,229,233,235,238,241,243,245,246,248,252,254,255,264,268,273,274,275,279,280,281,286,287,289,291,293,295,297,300,302,306,311,315,319,320,322,324,326,327,334,338,341,342,348,349,351,352,356,358,360,363,369,370,372,374,377,378,384,386,388,393,394,396,398,400,403,404,405,407,410,413,414,420,422,425,427,428,435,440,443,445,447,448,450,452,454,459,461,464,465,466,468,469,471,473,475,476,478,480,481,486,487,489,490,492,493,498,501,502,507,514,516,518,522,523,529,536,542,546,548,551,552,554,557,559,561,563,566,569,570,571,574,575,581,582,584,586,587,590,591,593,600,603,605,607,610,612,613,616,618,620,622,623,627,629,633,6

“some row.names duplicated: 5,8,9,13,15,17,18,20,22,24,25,27,28,31,33,36,37,41,43,45,47,48,51,57,58,60,66,68,69,72,73,75,76,79,80,82,83,87,91,93,94,96,99,104,105,106,109,110,111,112,114,116,117,120,123,124,131,132,133,140,143,145,146,147,152,159,163,166,167,173,178,179,186,192,194,196,199,205,207,209,212,216,219,220,223,224,226,229,236,238,241,242,244,246,250,253,256,258,261,263,264,265,271,272,276,280,287,288,291,298,300,301,302,304,306,310,313,316,317,320,321,322,324,331,332,334,336,338,339,345,346,349,357,358,360,362,365,368,369,373,379,385,390,394,397,398,401,403,405,407,408,411,416,418,420,421,424,428,432,433,436,438,439,443,444,445,448,454,455,459,461,463,465,473,475,478,485,496,498,501,504,510,512,515,516,518,522,525,526,527,530,532,533,535,538,539,547,548,553,556,558,560,561,563,565,568,569,570,572,575,576,579,581,582,583,585,586,589,596,599,600,602,603,608,610,612,616,620,623,625,626,629,631,632,635,639,640,646,647,649,656,663,666,670,672,677,681,683,686,687,688,689,693,694,69

“some row.names duplicated: 2,9,12,13,15,16,18,19,21,26,28,33,35,41,43,47,51,53,56,58,60,65,66,68,71,72,77,79,80,82,83,86,89,94,95,98,104,107,108,109,112,114,119,124,126,129,130,132,134,136,137,138,141,142,143,145,146,148,152,154,155,157,160,164,165,170,171,173,175,178,179,181,184,186,187,190,192,196,197,198,202,208,210,212,214,216,217,218,220,222,224,226,230,233,235,237,241,243,245,247,248,249,250,253,255,256,259,262,268,269,271,275,276,277,279,282,283,287,289,292,294,296,297,298,303,304,309,312,313,317,318,319,324,331,337,344,347,352,355,358,360,362,366,370,372,373,376,380,381,385,389,397,399,400,404,406,408,413,414,417,421,431,432,433,437,443,445,448,451,452,461,462,464,465,467,469,472,473,476,481,485,488,491,492,494,502,505,506,508,509,510,514,515,516,520,522,523,529,530,531,533,535,542,544,548,549,551,552,561,562,570,577,578,580,581,585,587,589,591,592,593,595,596,599,601,608,610,613,616,618,622,624,625,626,630,632,635,637,640,641,646,648,649,651,653,656,658,659,663,665,667,668,67

“some row.names duplicated: 2,3,8,10,12,14,15,16,19,20,22,23,25,27,32,33,37,38,40,41,42,43,45,49,51,52,53,56,59,64,66,69,74,76,78,82,86,87,88,90,91,93,95,97,99,104,113,119,121,123,124,129,135,138,139,141,143,147,149,152,154,155,160,162,164,166,168,169,173,174,175,177,180,181,189,190,192,193,195,197,200,205,207,210,214,217,218,221,224,225,231,236,238,241,242,243,246,250,251,253,258,260,262,265,266,267,270,273,275,277,279,280,281,282,283,289,291,293,295,297,298,301,302,304,305,306,310,312,313,319,322,326,330,332,334,336,339,340,341,343,345,346,348,349,351,354,358,361,362,363,365,366,368,370,375,377,380,383,385,392,393,396,398,399,400,402,403,404,407,411,413,415,417,423,424,427,430,433,435,438,441,442,446,447,453,458,463,464,467,471,473,478,481,486,489,493,494,499,501,502,511,513,515,518,520,521,524,530,535,539,548,549,550,554,558,560,561,564,565,567,572,573,575,578,584,587,592,594,595,596,599,601,602,603,609,620,621,623,624,628,629,634,636,637,639,641,642,644,645,650,652,654,658,662,666,

“some row.names duplicated: 2,4,5,20,21,23,32,33,38,40,41,43,44,48,51,53,55,57,61,65,67,68,72,74,76,77,79,80,82,86,89,91,92,94,105,110,115,119,120,122,125,129,132,135,137,138,143,145,147,152,154,156,159,161,163,166,167,168,171,172,173,174,177,178,179,181,184,192,194,197,202,203,205,213,214,216,218,223,227,228,232,236,238,240,243,244,246,248,256,257,260,261,264,266,267,269,273,274,276,278,282,284,285,288,296,297,298,299,301,302,303,305,306,314,319,320,325,326,330,331,337,338,340,343,350,351,353,360,361,362,364,370,371,373,375,377,379,380,385,387,390,391,392,394,398,400,402,403,404,408,409,411,421,422,425,428,431,433,435,437,438,440,442,445,453,457,458,459,461,462,466,471,476,479,480,487,488,492,493,495,496,497,500,501,508,510,513,516,520,522,524,525,528,536,538,539,543,545,547,548,557,558,560,561,563,566,567,569,571,574,575,577,580,586,590,593,596,601,602,603,605,608,609,611,613,616,620,621,625,626,627,630,631,638,639,641,642,645,646,649,650,654,655,657,660,663,665,668,672,678,681,683,6

“some row.names duplicated: 3,4,7,11,12,14,22,24,26,27,29,32,36,37,39,41,42,46,48,49,53,62,64,65,67,69,70,72,74,79,81,82,83,87,91,95,97,101,103,109,110,111,113,115,116,119,121,123,124,127,129,132,134,137,139,140,144,147,148,150,157,161,163,165,167,168,169,171,172,173,176,178,181,182,185,186,189,190,193,194,196,201,203,204,206,207,214,220,222,230,231,234,236,237,239,243,245,247,249,263,265,266,267,274,277,283,287,289,293,294,302,304,306,307,309,310,312,313,314,317,318,319,322,324,325,327,330,337,341,346,353,357,358,362,364,372,377,380,381,383,387,391,394,397,401,404,409,410,413,415,416,423,424,425,427,428,430,437,438,441,442,447,449,453,454,459,460,463,464,466,469,471,472,474,479,481,483,488,493,495,496,499,502,505,506,510,511,515,520,522,523,524,526,529,530,535,537,545,546,548,552,559,562,564,566,568,570,572,573,575,579,582,583,588,591,593,597,598,600,601,605,606,615,623,628,631,633,635,637,641,642,644,649,651,654,655,657,658,659,663,664,665,666,668,673,675,680,683,687,690,692,694,696,

“some row.names duplicated: 6,8,14,16,20,22,23,31,33,38,44,46,47,49,50,52,54,57,59,63,64,67,68,69,71,73,74,76,79,81,82,85,97,99,102,103,109,115,119,123,127,130,133,137,138,140,141,145,147,148,149,151,152,156,158,160,163,165,169,173,175,177,183,184,186,187,189,194,195,197,200,201,205,208,209,212,213,217,218,221,223,225,230,237,239,244,245,248,249,251,253,259,261,262,264,265,268,269,270,274,279,283,284,286,287,292,295,302,304,307,310,312,316,320,322,326,328,331,338,340,349,351,355,362,363,365,368,373,374,378,382,383,384,387,388,389,392,393,397,398,399,401,405,406,409,410,412,413,415,417,420,425,431,433,434,439,440,443,446,452,454,456,460,461,466,468,472,482,483,487,488,490,494,496,499,502,505,507,511,512,515,518,520,522,523,525,528,532,535,536,538,540,542,543,545,548,549,553,560,567,569,570,574,581,582,590,593,600,601,603,607,612,614,620,624,628,631,633,635,638,644,649,650,651,654,658,659,661,662,663,667,669,670,674,678,682,686,690,693,695,699,704,705,706,709,710,713,716,720,721,723,725,

“some row.names duplicated: 2,3,5,9,15,16,17,19,23,27,29,31,36,39,40,41,42,44,46,50,52,54,57,58,60,63,64,66,68,74,75,77,78,81,85,88,89,91,93,95,99,105,106,107,108,110,111,114,118,121,123,126,129,131,134,136,138,139,140,145,148,149,154,156,159,160,163,166,168,170,176,177,182,185,187,192,194,195,197,203,205,207,209,212,217,219,222,224,227,228,231,232,236,238,239,242,244,245,252,253,254,256,257,259,261,266,273,275,277,279,282,284,286,287,291,293,294,295,297,298,300,301,303,308,310,312,316,319,321,322,325,330,331,333,336,338,343,348,350,357,360,362,366,367,371,372,373,374,376,381,389,391,393,399,411,420,425,428,432,434,436,437,439,443,445,446,451,453,456,458,459,461,467,469,472,475,477,479,480,482,485,488,489,491,493,495,505,512,513,515,518,519,520,522,523,524,528,529,531,532,535,538,539,544,546,548,550,553,555,556,560,562,563,564,566,569,573,576,579,584,585,592,593,603,604,609,614,618,622,625,627,630,633,634,639,641,645,649,656,669,671,674,676,678,679,682,687,688,691,692,697,701,704,705,7

“some row.names duplicated: 2,4,8,9,22,26,32,34,36,37,39,44,47,48,49,50,52,55,57,58,60,61,63,66,68,70,76,77,79,80,84,86,87,91,95,96,98,99,101,102,105,107,111,115,117,118,123,124,126,128,129,131,133,136,138,142,144,147,149,150,152,154,158,162,168,171,173,174,181,183,186,192,195,196,198,200,204,206,210,214,216,217,220,226,227,229,230,232,237,238,239,244,245,247,252,254,264,269,274,275,276,281,286,288,291,295,296,299,305,306,307,310,320,322,323,326,330,335,342,346,347,351,353,358,359,362,363,364,368,370,372,374,381,384,391,392,393,394,396,402,403,404,405,412,414,415,418,425,426,427,428,429,430,432,434,437,439,440,444,446,448,451,455,464,465,466,467,470,471,473,477,479,482,483,486,490,491,492,505,509,510,511,516,519,526,527,529,533,534,538,539,540,543,549,550,552,555,558,559,561,569,574,575,576,580,584,586,591,598,600,603,607,609,610,612,621,624,626,629,631,633,634,635,636,639,641,645,649,650,651,653,655,658,661,665,667,669,671,672,676,677,679,682,683,685,686,688,690,691,693,694,698,702,70

“some row.names duplicated: 2,4,6,9,14,17,22,23,29,36,38,39,42,44,46,50,53,54,55,58,59,64,66,70,74,76,80,81,87,89,95,97,100,106,108,109,110,113,119,122,124,126,129,130,134,137,140,141,144,149,150,152,160,166,169,176,178,179,181,183,185,186,190,191,193,199,200,202,203,205,209,212,213,218,221,224,228,235,238,240,241,242,246,247,250,253,256,257,259,261,268,271,272,273,274,277,285,291,298,301,306,316,319,320,323,325,326,331,332,335,336,338,340,347,352,354,358,360,361,365,367,375,377,380,390,392,395,396,398,399,405,410,413,414,419,420,422,423,427,432,433,434,435,441,443,444,450,453,455,457,461,462,463,465,466,467,470,471,479,480,482,484,489,490,494,495,503,504,507,509,515,520,521,522,525,528,529,530,532,534,536,537,538,541,543,544,546,547,548,551,555,560,564,567,569,570,575,576,577,579,582,583,586,588,590,593,598,605,608,609,616,620,621,622,627,628,631,634,639,640,642,643,645,646,649,651,655,659,664,665,666,668,670,675,678,682,683,685,689,690,691,694,696,698,700,702,703,706,707,719,722,727,

“some row.names duplicated: 2,3,7,8,13,14,15,18,22,23,27,32,39,41,45,48,57,60,61,64,69,71,73,75,87,89,92,96,98,100,102,105,106,108,112,113,115,120,122,123,124,128,129,131,132,134,137,140,142,144,160,162,163,165,168,169,172,173,178,181,185,188,190,191,194,197,198,211,212,214,217,218,220,221,224,226,227,229,230,235,238,247,249,250,254,263,268,272,273,277,280,285,290,292,293,295,302,305,307,308,309,310,311,315,317,319,326,328,331,332,337,340,343,347,349,350,353,357,358,360,368,370,371,373,376,383,385,387,388,390,393,395,397,400,403,406,409,412,414,416,421,422,423,425,427,433,434,437,438,439,441,443,445,446,449,461,467,468,470,472,475,476,478,480,483,486,491,497,501,505,508,510,512,513,514,515,517,529,531,536,538,539,544,545,548,549,550,553,554,556,557,561,563,568,570,573,576,579,582,584,587,590,592,593,594,597,599,608,613,616,617,618,620,622,623,624,625,627,629,632,633,640,642,643,645,648,650,652,656,658,662,663,665,667,668,672,676,679,681,683,684,689,691,692,697,700,706,708,714,716,717,7

“some row.names duplicated: 2,6,8,9,10,11,19,22,25,26,28,30,41,43,44,46,47,51,53,57,58,59,65,68,71,72,74,78,82,87,89,90,94,97,102,103,106,108,109,111,115,118,120,121,122,126,131,133,137,139,141,143,147,149,152,154,163,168,170,175,177,179,181,182,184,190,191,193,196,197,200,204,206,207,209,215,217,220,223,225,229,230,236,237,239,242,246,248,254,256,258,262,265,266,269,271,272,277,279,286,289,292,293,295,296,298,301,304,306,309,311,315,318,320,324,325,329,330,332,334,335,339,340,344,349,352,356,359,363,365,367,370,372,376,377,378,380,381,384,385,387,391,393,400,402,403,405,406,412,413,416,422,423,427,428,432,434,437,438,442,444,447,450,456,461,465,466,467,468,470,471,473,476,478,480,484,486,489,492,503,505,509,511,515,516,518,522,533,534,535,538,539,542,549,552,554,557,558,561,568,569,571,574,576,579,581,584,586,587,589,590,594,595,601,603,607,608,609,610,612,613,616,617,624,628,631,635,636,644,646,648,650,651,652,654,656,659,663,664,668,669,671,675,677,678,684,687,691,696,703,705,706,71

“some row.names duplicated: 2,4,5,6,8,9,16,18,21,22,24,27,36,44,46,48,49,51,53,55,59,60,61,68,69,71,72,73,75,77,78,80,81,83,84,87,89,91,95,100,106,108,111,116,118,120,122,123,124,125,132,133,135,138,139,142,145,148,149,151,154,157,158,160,161,162,165,178,185,187,191,194,195,207,211,213,217,218,219,223,224,227,236,237,240,241,244,247,248,253,255,256,258,259,260,262,264,267,271,275,276,278,279,283,286,291,292,294,299,301,302,304,306,311,313,316,324,327,331,333,335,336,340,344,345,346,350,351,355,361,363,364,366,373,374,377,378,379,382,386,390,393,394,397,400,401,403,405,408,409,413,420,424,427,429,430,431,434,437,439,441,446,450,451,454,456,461,464,466,467,481,484,487,490,492,495,496,502,504,505,509,511,513,515,516,519,520,527,535,542,543,544,547,548,552,554,556,559,560,562,575,580,582,583,590,602,603,607,609,611,613,614,615,617,619,621,622,627,633,634,635,637,642,644,645,647,649,654,656,658,661,664,665,668,669,670,671,672,674,677,682,686,687,689,694,698,699,701,707,710,712,713,714,718,7

“some row.names duplicated: 3,8,10,13,17,19,20,21,24,25,32,35,36,39,41,46,48,49,51,55,56,59,64,69,75,76,78,79,83,85,88,91,94,96,97,99,103,105,108,109,111,116,122,123,131,135,137,138,144,146,147,149,152,155,157,158,159,163,166,169,172,174,175,181,182,185,186,188,189,191,195,198,201,207,208,210,211,213,216,221,225,227,229,232,233,236,237,241,242,245,247,255,257,259,260,262,263,266,274,275,279,285,288,289,294,295,297,298,299,302,309,312,316,323,324,325,328,329,331,333,338,340,342,343,347,348,349,357,360,366,369,372,373,375,378,379,382,391,392,394,397,398,401,402,404,407,410,412,419,420,422,423,424,426,427,429,430,432,434,436,439,440,442,445,446,448,450,451,454,455,456,461,462,464,468,472,473,477,484,485,489,491,492,493,495,496,498,499,504,505,506,509,511,513,514,515,520,525,528,530,534,535,538,540,544,548,551,556,562,564,565,567,568,571,572,575,577,584,588,590,599,600,601,602,604,607,609,612,616,618,624,626,627,628,629,630,632,637,639,640,644,646,647,652,657,658,660,663,665,667,672,674,67

“some row.names duplicated: 2,5,6,8,9,12,14,22,26,28,30,32,33,38,40,41,42,48,50,52,53,55,57,58,61,69,74,75,79,81,83,85,89,92,97,98,101,102,104,109,111,115,117,118,120,122,123,126,129,140,143,147,154,159,161,164,166,168,170,176,177,179,181,184,188,189,192,202,204,205,209,211,215,216,217,227,231,233,240,241,243,246,247,249,254,255,259,261,267,271,273,275,276,278,279,281,285,293,294,300,302,303,309,313,316,318,320,326,328,332,336,338,340,343,344,346,348,349,351,352,353,355,357,359,364,366,368,369,373,376,378,379,383,385,387,390,391,395,396,398,399,401,404,411,414,416,419,420,423,424,426,429,430,432,433,437,440,443,444,448,451,455,458,459,463,469,473,475,479,481,482,485,487,489,490,493,497,499,503,507,511,513,516,518,524,526,529,537,538,543,545,546,548,552,556,559,560,564,565,569,571,572,574,576,577,578,580,584,586,587,589,594,601,602,603,609,611,613,616,621,623,626,628,631,633,634,636,638,645,648,651,655,657,659,662,667,669,670,672,673,675,679,682,690,692,694,695,696,697,698,704,706,707,7

“some row.names duplicated: 2,5,6,8,9,12,14,15,17,20,21,24,25,28,31,37,40,41,45,48,50,52,54,57,60,62,64,66,69,71,74,76,78,79,81,84,93,94,96,97,99,107,109,110,113,116,119,122,124,127,131,133,134,135,139,140,143,144,147,149,150,157,158,164,165,167,170,171,173,174,179,181,183,184,185,187,190,193,197,199,201,202,208,211,214,215,217,219,221,226,228,230,232,235,236,237,250,251,253,259,260,261,263,264,268,274,278,281,283,287,288,290,291,293,294,296,298,301,303,306,309,315,319,320,322,323,326,331,332,334,336,337,340,342,347,349,350,354,355,358,360,362,363,365,366,370,373,374,375,377,382,383,387,388,392,395,396,402,404,406,408,409,412,413,415,417,419,420,422,423,425,429,431,433,435,437,439,447,449,454,457,458,470,472,476,478,483,485,486,488,492,494,495,496,497,504,505,509,512,518,521,523,527,528,529,531,536,538,540,543,544,546,547,548,550,551,555,556,559,560,562,564,568,569,570,574,577,583,584,585,587,591,595,598,602,604,606,607,612,615,619,621,625,631,633,635,638,640,646,648,650,653,655,657,65

“some row.names duplicated: 4,6,7,8,9,10,13,15,16,22,24,27,28,31,33,35,36,38,43,46,47,49,52,53,55,56,61,68,69,82,85,87,89,91,93,95,97,100,104,106,111,118,119,128,133,135,136,140,143,146,152,153,157,159,160,162,164,167,168,170,172,174,178,179,185,191,195,197,198,205,210,211,212,214,218,221,223,226,229,232,236,237,247,250,251,254,255,262,263,271,272,273,276,282,287,290,293,304,305,306,308,311,312,313,314,316,320,324,328,329,330,334,335,336,338,339,344,346,348,350,358,360,362,363,364,367,370,375,376,378,380,382,386,391,393,396,402,404,406,408,409,411,414,416,417,420,424,426,428,430,431,432,435,439,441,442,445,446,449,450,454,456,460,464,466,468,473,475,482,485,487,490,496,498,501,503,504,506,510,514,519,522,525,532,533,534,539,543,546,550,551,558,560,563,564,566,567,569,571,574,575,577,578,579,581,584,586,587,588,590,592,593,595,596,602,603,608,612,614,618,619,621,624,626,628,631,634,646,652,653,654,655,658,661,663,666,669,672,675,679,681,682,684,685,687,688,690,693,700,706,709,710,713,71

“some row.names duplicated: 2,4,7,9,11,12,16,17,19,20,23,25,27,32,33,35,36,38,40,41,43,45,49,53,55,56,64,66,67,70,73,76,78,82,84,86,90,95,97,100,105,107,109,111,114,115,120,122,124,125,129,131,132,134,135,136,139,143,145,149,156,160,163,169,174,175,177,180,185,187,194,195,197,199,204,205,206,208,210,214,218,219,221,222,223,228,230,233,236,237,240,243,245,250,252,253,262,264,266,270,273,274,275,278,279,282,285,290,297,298,302,305,306,307,309,311,314,316,318,321,322,329,336,338,340,341,343,347,349,350,352,361,365,366,367,369,372,374,383,384,385,389,390,391,395,396,403,404,407,408,411,413,414,416,420,424,427,429,430,432,440,443,445,449,451,452,454,457,460,462,466,467,471,474,477,478,479,482,485,490,491,492,495,497,502,504,509,510,514,516,523,524,528,530,533,534,535,540,541,542,546,547,548,550,554,558,560,564,565,566,569,579,580,583,584,587,590,592,593,604,605,609,612,615,616,625,626,628,629,630,632,635,638,641,642,643,644,651,654,660,662,664,669,676,681,683,686,688,690,691,693,697,702,705

“some row.names duplicated: 3,6,8,9,14,17,19,21,22,24,26,29,31,33,36,45,47,48,50,52,56,61,63,67,69,70,72,74,75,78,82,84,88,89,97,99,101,103,106,111,114,119,123,125,127,133,137,138,142,146,148,151,153,163,165,167,169,171,172,177,183,189,194,197,199,200,206,207,209,210,214,216,217,219,220,223,224,225,227,229,232,234,236,237,240,244,248,252,253,256,257,259,261,262,263,266,272,276,278,280,283,287,293,296,302,304,307,310,311,313,314,321,325,328,330,331,337,341,345,346,348,350,351,352,355,357,362,363,364,366,367,368,373,374,376,377,379,381,383,385,388,392,398,402,407,409,414,418,422,427,431,432,433,435,437,438,442,445,449,452,457,460,462,466,469,475,480,485,486,487,488,495,499,501,503,505,509,512,513,515,517,518,520,521,524,525,527,529,531,533,534,536,537,539,541,546,549,550,553,556,558,563,565,566,569,574,576,577,578,582,583,584,591,592,595,596,600,603,605,607,610,613,614,616,618,620,623,625,630,633,634,640,644,654,657,662,666,673,676,677,679,680,682,683,684,686,688,691,696,701,704,706,710,

“some row.names duplicated: 2,5,6,10,11,15,16,19,21,23,26,28,33,37,38,40,41,44,45,47,49,51,53,59,63,64,67,71,72,74,75,78,79,82,83,85,87,89,91,92,97,99,101,103,105,107,108,113,118,122,123,128,129,133,134,136,139,147,148,152,156,158,159,163,165,166,168,169,175,178,179,183,188,190,193,194,199,202,209,213,214,220,222,227,229,231,233,239,241,247,248,251,257,258,260,261,263,264,266,267,270,273,278,281,284,291,293,300,301,305,306,307,310,312,313,319,322,324,325,326,327,332,333,336,338,344,346,348,350,352,353,355,359,363,366,368,369,370,372,375,377,379,380,382,388,392,395,396,398,399,401,402,412,413,415,417,419,421,422,424,429,430,433,436,438,442,443,454,456,459,462,465,468,470,473,475,477,482,484,485,486,489,490,494,495,498,500,502,507,508,511,513,515,516,521,523,530,533,534,541,545,546,548,550,551,553,556,560,564,568,569,571,572,575,577,578,583,586,600,604,606,608,609,611,612,613,615,616,617,619,620,621,623,627,628,630,632,633,635,641,645,648,649,651,652,657,659,661,662,668,672,675,679,680,6

Something is wrong; all the Accuracy metric values are missing:
    Accuracy       Kappa    
 Min.   : NA   Min.   : NA  
 1st Qu.: NA   1st Qu.: NA  
 Median : NA   Median : NA  
 Mean   :NaN   Mean   :NaN  
 3rd Qu.: NA   3rd Qu.: NA  
 Max.   : NA   Max.   : NA  
 NA's   :1     NA's   :1    


ERROR: Error: Stopping


In [29]:
svmbag

ERROR: Error in eval(expr, envir, enclos): object 'svmbag' not found


Error :(

## boosting

Similar to bagging, boosting uses ensembles of models trained on resampled data and a vote to determine the final prediction.
+ Resampled datasets in boosting are constructed specifically to generate complementary learners
+ Rather than giving each learner an equal vote, boosting gives each learner's vote a weight based on its past performance

Beginning from an unweighted dataset, the first classifier attempts to model the outcome. Examples that the classifier predicted correctly will be less likely to appear in the training dataset for the following classifier, and conversely, the difficult-to-classify examples will appear more frequently. As additional rounds of weak learners are added, they are trained on data with successively more difficult examples.

Boosting is most commonly used for decision trees.

In [32]:
library(adabag)

Loading required package: rpart
Loading required package: mlbench

Attaching package: ‘adabag’

The following object is masked from ‘package:ipred’:

    bagging



In [33]:
set.seed(300)
m_adaboost = boosting(default ~ ., data = credit)

In [34]:
p_adaboost = predict(m_adaboost, credit)

In [35]:
head(p_adaboost$class)

[1] "no"  "yes" "no"  "no"  "yes" "no"

In [36]:
p_adaboost$confusion

               Observed Class
Predicted Class  no yes
            no  700   0
            yes   0 300

In [37]:
set.seed(300)
adaboost_cv = boosting.cv(default ~ ., data = credit)

i:  1 Thu Jun  8 11:45:46 2017 
i:  2 Thu Jun  8 11:45:52 2017 
i:  3 Thu Jun  8 11:45:59 2017 
i:  4 Thu Jun  8 11:46:05 2017 
i:  5 Thu Jun  8 11:46:11 2017 
i:  6 Thu Jun  8 11:46:18 2017 
i:  7 Thu Jun  8 11:46:24 2017 
i:  8 Thu Jun  8 11:46:30 2017 
i:  9 Thu Jun  8 11:46:37 2017 
i:  10 Thu Jun  8 11:46:43 2017 


In [38]:
adaboost_cv$confusion

               Observed Class
Predicted Class  no yes
            no  594 151
            yes 106 149

In [39]:
library(vcd)

Loading required package: grid


In [40]:
Kappa(adaboost_cv$confusion)

            value    ASE     z  Pr(>|z|)
Unweighted 0.3607 0.0323 11.17 5.914e-29
Weighted   0.3607 0.0323 11.17 5.914e-29

## random forests

Random Forests combine versatility and power into a single machine learning approach. As the ensemble uses only a small, random portion of the full feature set, random forests can handle extremely large datasets, where the so-called "curse of dimensionality" might cause other models to fail.

+ An all-purpose model that performs well on most problems
+ Can handle noisy or missing data as well as categorical or continuous features
+ Selects only the most important features
+ Can be used on data with an extremely large number of features or examples
- Unline a decision tree, the model is not easily interpretable
- May require some work to tune the model to the data

In [41]:
# randomForests(train, class, ntree, mtry)
# train = data frame containing training data
# class = factor vector with the class for each row in the training data
# ntree = number of trees to grow
# mtry = number of features to randomly select at each split

In [45]:
library('randomForest')

randomForest 4.6-12
Type rfNews() to see new features/changes/bug fixes.

Attaching package: ‘randomForest’

The following object is masked from ‘package:ggplot2’:

    margin



In [48]:
set.seed(300)
rf = randomForest(default ~ ., data = credit)

In [49]:
rf


Call:
 randomForest(formula = default ~ ., data = credit) 
               Type of random forest: classification
                     Number of trees: 500
No. of variables tried at each split: 4

        OOB estimate of  error rate: 23.8%
Confusion matrix:
     no yes class.error
no  640  60  0.08571429
yes 178 122  0.59333333

Trees included: 500, tried 4 variables at each split.

out-of-bag error, which unlike the resubstitution error, is an unbiased estimate of the test set error. It should be a fairly reasonable estimate of future perfomance.

### evaluating RF peformance

In [50]:
ctrl = trainControl(method = 'repeatedcv', number = 10, repeats = 10)

In [51]:
grid_rf = expand.grid(.mtry = c(2, 4, 8, 16))

In [53]:
set.seed(300)
m_rf = train(default ~ ., data = credit, method = 'rf',
            metric = "Kappa", trControl = ctrl, tuneGrid = grid_rf)

In [54]:
grid_c50 = expand.grid(.model = 'tree',
                      .trials = c(10, 20, 30, 40),
                      .winnow = 'FALSE')

In [55]:
set.seed(300)
m_c50 = train(default ~ ., data = credit, method = 'C5.0',
             metric = 'Kappa', trControl = ctrl, tuneGrid = grid_c50)

Warning message in Ops.factor(x$winnow):
“‘!’ not meaningful for factors”

In [56]:
m_rf

Random Forest 

1000 samples
  16 predictor
   2 classes: 'no', 'yes' 

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 10 times) 
Summary of sample sizes: 900, 900, 900, 900, 900, 900, ... 
Resampling results across tuning parameters:

  mtry  Accuracy  Kappa    
   2    0.7256    0.1311283
   4    0.7476    0.2878470
   8    0.7519    0.3346061
  16    0.7557    0.3618152

Kappa was used to select the optimal model using  the largest value.
The final value used for the model was mtry = 16.

In [57]:
m_c50

C5.0 

1000 samples
  16 predictor
   2 classes: 'no', 'yes' 

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 10 times) 
Summary of sample sizes: 900, 900, 900, 900, 900, 900, ... 
Resampling results across tuning parameters:

  trials  Accuracy  Kappa    
  10      0.7325    0.3215655
  20      0.7343    0.3268052
  30      0.7381    0.3343137
  40      0.7388    0.3335082

Tuning parameter 'model' was held constant at a value of tree
Tuning
 parameter 'winnow' was held constant at a value of FALSE
Kappa was used to select the optimal model using  the largest value.
The final values used for the model were trials = 30, model = tree and winnow
 = FALSE.

With a kappa of about .361, the random forest model with mtry = 16 was the winner among these eight models. It was higher than the best C5.0 decision tree, which had a kappa of about .334, and slightly higher than the AdaBoost.M1 model with a kappa of about .36. Based on these results we would submit the random forest as our final model.